### PR TITLE
Reorder Items

### LINKED ISSUE #40: Reorder items
I can't reorder my to-do list items

### PR DESCRIPTION
**Pull request by tycebrown**
Major "code smell":

My reordering algorithm is really simple: it just gets all the items that will need their order changed, updates them and puts the updated items to the database. Because of the last step I initially decided to make the reordering function asynchronous (since our database method are, at the current moment, all asynchronous). However, this introduced an inconsistently appearing visual bug (in the screen recording, at about 14 seconds, when I move the item "4"):

Demonstration

I realized this bug is probably due to the fact that the `onReorder` callback of the [reorderables](https://pub.dev/packages/reorderables) is called synchronously (so the reordering work is not "awaited" to be completed before the reorderable rebuilds).

I ended up making the reordering work synchronous and changing the last step of the algorightm to call (but not await) the `DatabaseManager.putItems` method. I had to do a little more weaving to get this to work.

All this to say, I looked at the code and think it would be more convenient if we made all (or most) of our database methods synchronous. Should we do this? Maybe in another issue? (Also, should I have already done it in this branch?)

Closes #40

### CODE REVIEW

**Comment by Pertempto** [comment]
It generally is a good idea to stick with async code. It might make sense to do the refactor in this case. That's a major change though, so we should not bog down this MR. Remember, smaller MRs are much better than bigger MRs.

Sometimes it even makes sense to use multiple MRs to complete one issue, breaking the changes into separate chunks that go in each MR. This makes code review much easier and quicker.

**Comment by Pertempto** [comment]
Don't worry about changing any of the code until I get a chance to look at it

**Comment by Pertempto** [comment]
Hmmm something isn't working quite right

I start with a two item list
image
I drag "Test" to the top and get this
image

This error is in the logs:
```
The following assertion was thrown while routing a pointer event:
'package:lists/model/list_model.dart': Failed assertion: line 93 pos 12: 'oldOrder < items.length &&
oldOrder >= 0': is not true.
```

Can you look into it? I'll take a look at the code after #41 is completed.

**Comment by tycebrown** [comment]
Ah, I think I found the problem. The order property for each item is loaded from isar, but any items created in previous versions of the app won't have their order stored in the database. If you create a new list and add items to it, you should be able to reorder them.

It would also be possible to write some quick and dirty code which would assign an order to items that don't have it.

**Comment by Pertempto** [comment]
Many files have merge conflicts.

**Comment by Pertempto** [comment]
The functionality works great! :smile:

**Comment by Pertempto** [comment]
I will review the code after you resolve the conflicts.

**Comment by tycebrown** [comment]
Merged! :tada:

File: lib/view/list_widget.dart [hunk 1]
```diff
@@ -63,24 +65,62 @@ class _ListWidgetState extends State<ListWidget> {
...
+      unCheckedItemWidgetsSliver,
+      if (checkedItems.isNotEmpty) dividerSliver,
+      checkedItemWidgetsSliver
+    ]);
   }
 
-  Iterable<Widget> _buildItemWidgets({required Iterable<Item> fromItems}) =>
-      fromItems.map((item) => ItemWidget(item, onDelete: () async =>
+  List<Widget> _buildItemWidgets({required Iterable<Item> fromItems, bool tappable = true}) =>
+      fromItems.map((item) => ItemWidget(item,tappable: tappable,  onDelete: () async =>
```
Comment (Pertempto) [inline lib/view/list_widget.dart]:
```suggestion
      fromItems.map((item) => ItemWidget(item, tappable: tappable, onDelete: () async =>
```

**Comment by Pertempto** [comment]
There is no need for your `order` logic to be so strict about `order` being between 0 and the length of the list.

You can simplify the logic by removing the restriction. `order` can be any integer.

**Comment by Pertempto** [comment]
@tycebrown I created some changes that remove the need for `reorderItems` to be synchronous.

https://github.com/Pertempto/Lists/pull/54 will merge into your branch. Please review it and merge if it works well on your end.

**Comment by Pertempto** [comment]
I'm not sure why checkable items show up like this on Android

**Review by Pertempto — COMMENTED** [review]
Please comment and/or merge #54, then I can continue with this one.

**Comment by Pertempto** [comment]
All that is left here is fixing the visual glitch with checkbox items.

**Comment by tycebrown** [comment]
> All that is left here is fixing the visual glitch with checkbox items.

Hmm... they show up fine in linux. I'll test on an android device...

**Comment by tycebrown** [comment]
I can reproduce on android. Looking into it...

**Review by Pertempto — APPROVED** [review]
@tycebrown I have a fix for the visual glitch in my branch for #43. You can merge this when you want.

**Pull request merged**